src/TestSessionEnvironment.php

```diff
@@ -282,54 +276,93 @@ public function applyState($state)
-                $dbExists = DB::get_conn()->databaseExists($dbName);
-            } else {
-                $dbExists = false;
+        $databaseConfig = DB::getConfig();
+        if (!$dbExists && $dbCreate) {
+            $this->oldDatabaseName = $databaseConfig['database'];
+            // Create a new one with a randomized name
+            $tempDB = new TempDatabase();
+            $dbName = $tempDB->build();
+
+            $state->database = $dbName; // In case it's changed by the call to SapphireTest::create_temp_db();
+
+            // Set existing one, assumes it already has been created
+            $prefix = Environment::getEnv('SS_DATABASE_PREFIX') ?: 'ss_';
+            $pattern = strtolower(sprintf('#^%stmpdb.*#', preg_quote($prefix, '#')));
+            if (!preg_match($pattern, $dbName)) {
+                throw new InvalidArgumentException("Invalid database name format");
             }
+        }
 
-            if (!$dbExists) {
-                // Create a new one with a randomized name
-                $tempDB = new TempDatabase();
-                $dbName = $tempDB->build();
-
-                $state->database = $dbName; // In case it's changed by the call to SapphireTest::create_temp_db();
+        if ($dbExists || $dbCreate) {
+            // Connect to the new database, overwriting the old DB connection (if any)
+            $databaseConfig['database'] = $dbName; // Instead of calling DB::set_alternative_db_name();
+            DB::connect($databaseConfig);
+            TestSessionState::create()->write();  // initialize the session state
+        }
 
-                // Set existing one, assumes it already has been created
-                $prefix = Environment::getEnv('SS_DATABASE_PREFIX') ?: 'ss_';
-                $pattern = strtolower(sprintf('#^%stmpdb.*#', preg_quote($prefix, '#')));
-                if (!preg_match($pattern, $dbName)) {
-                    throw new InvalidArgumentException("Invalid database name format");
-                }
+        return $state;
+    }
 
-                $this->oldDatabaseName = $databaseConfig['database'];
-                $databaseConfig['database'] = $dbName; // Instead of calling DB::set_alternative_db_name();
+    /**
+     * Takes a state object and applies environment transformations to current application environment
+     *
+     * Does not persist any changes to the the state object - @see saveState for persistence
+     *
+     * Assumes the database has already been created in startTestSession(), as this method can be called from
+     * _config.php where we don't yet have a DB connection.
+     *
+     * You can extend this by creating an Extension object and implementing either onBeforeApplyState() or
+     * onAfterApplyState() to add your own test state handling in.
+     *
+     * @param mixed $state
+     * @throws LogicException
+     * @throws InvalidArgumentException
+     * @return mixed $state
+     */
+    public function applyState($state)
```

Comment:
Unfortunately, `applyState` is a public method and we cannot change its behaviour without introducing breaking changes.
What do you say if we rename this method to `setState` and then implement the `applyState` original behaviour just running the new API internally?
 - `createDatabase`
 - `mergeWithExistingState`
 - `setState`

In that case `applyState` behaviour would stay as is, so we keep backward compatibility and in the same time we introduce new API (and use it in start test session session instead of apply state?)

Reply:
I feel like backwards compatibility with this will be quite challenging. The current state has quite a lot of mixed logic. Create test session sets some things but then relies on the middleware and apply state to fill in the rest. I'd be more in favour of a major version bump as it seems a lot cleaner (and less work to re-implement the existing behaviour).

Reply:
Yeah, I just discussed with Ingo. I guess it's gonna be much less resistant and more productive to issue a new major version when we need it. Let's rebase this on master and break stuff! :)

Reply:
rebased - should I target master with the PR?

Reply:
yes please